File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/AccessoryAgentSample/src/com/hp/workpath/sample/accessoryagentsample/receiver/AccessoryReceiver.java
package com.hp.workpath.sample.accessoryagentsample.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.text.TextUtils;
import android.util.Base64;
import android.util.Log;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.SsdkUnsupportedException;
import com.hp.workpath.api.Workpath;
import com.hp.workpath.api.access.AccessService;
import com.hp.workpath.api.access.Principal;
import com.hp.workpath.api.accessory.RegistrationType;
import com.hp.workpath.api.accessory.hid.AccessoryService;
import com.hp.workpath.api.accessory.hid.EventCode;
import com.hp.workpath.api.accessory.hid.HIDAccessoryInfo;
import com.hp.workpath.api.accessory.hid.HIDInfo;
import com.hp.workpath.api.accessory.hid.HIDReport;
import com.hp.workpath.api.accessory.hid.HIDReportEventInfo;
import com.hp.workpath.sample.accessoryagentsample.Logger;
import com.hp.workpath.sample.accessoryagentsample.MainActivity;

import java.lang.ref.WeakReference;
import java.util.List;

public class AccessoryReceiver extends BroadcastReceiver {
    private static final String TAG = MainActivity.TAG + "Receiver";
    private static final String ACCESSORY_CONTEXT_CHANGE_ACTION = "com.hp.workpath.api.accessory.ACCESSORY_CONTEXT_CHANGE_ACTION";
    private static final String ACCESSORY_CHANGE_ACTION = "com.hp.workpath.api.accessory.ACCESSORY_CHANGE_ACTION";
    private WeakReference<Context> mContextRef;
    private Context appContext;
    private static boolean isInitializedSDK = false;

    @Override
    public void onReceive(Context context, Intent intent) {
        appContext = context.getApplicationContext();
        this.mContextRef = new WeakReference<>(appContext);

        initService(new Handler(), appContext);

        if(ACCESSORY_CONTEXT_CHANGE_ACTION.equals(intent.getAction())
                || ACCESSORY_CHANGE_ACTION.equals(intent.getAction())){
            final String action = intent.getAction();

            switch (action) {
                case ACCESSORY_CONTEXT_CHANGE_ACTION:
                    onContextChange(intent);
                    break;
                case ACCESSORY_CHANGE_ACTION:
                    onChange(intent);
                    break;
            }
        }
    }

    private void initService(Handler handler, Context context) {
        try {
            Log.i(TAG,"initService Workpath.getInstance().initialize");
            if(!isInitializedSDK) {
                Workpath.getInstance().initialize(mContextRef.get());
                isInitializedSDK = true;
            }
        } catch (SsdkUnsupportedException e) {
            Logger.showResult(null,e.getMessage());
        }
    }

    private void openAccessory(String accessoryContextId) {
        Log.d(TAG, "Accessory opened: " + accessoryContextId);
        Result result = new Result();
        try{
            AccessoryService.open(appContext, accessoryContextId, result);
            Logger.showResult(null, "AccessoryService.open", result);
        } catch(Exception e) {
            Log.e(TAG,e.getMessage(),e);
        }

        try{
            HIDInfo hidInfo = AccessoryService.getInfo(appContext,accessoryContextId, result);
            if (result.getCode() == Result.RESULT_OK) {
                Log.d(TAG, "HID feature report length : " + hidInfo.getFeatureReportLength());
                Log.d(TAG, "HID input report length : " + hidInfo.getInputReportLength());
                Log.d(TAG, "HID output report length : " + hidInfo.getOutputReportLength());
            }
        } catch(Exception e) {
            Log.e(TAG,e.getMessage(),e);
        }

        try{
            AccessoryService.startReading(appContext, accessoryContextId, result);
            Logger.showResult(null, "AccessoryService.startReading", result);
        } catch(Exception e) {
            Log.e(TAG,e.getMessage(),e);
        }
    }


    private void initiateSignIn() {
        Result result = new Result();
        Log.d(TAG, "initiateSignIn");
        AccessService.initiateSignIn(appContext, result);
        Logger.showResult(null, "AccessService.initiateSignIn", result);
    }

    private void signOut() {
        Result result = new Result();
        Log.d(TAG, "signOut");
        AccessService.signOut(appContext, result);
        Logger.showResult(null, "AccessService.signOut", result);
    }

    private Principal getCurrentPrincipal() {
        Result result = new Result();
        Log.d(TAG, "getCurrentPrincipal");
        Principal principal = AccessService.getCurrentPrincipal(appContext, result);
        Logger.showResult(null, "AccessService.getCurrentPrincipal", result);
        if (result.getCode() == Result.RESULT_OK) {
            return principal;
        } else {
            return null;
        }
    }

    public void onContextChange(Intent intent) {
        HIDAccessoryInfo hidAccessoryInfo = intent.getParcelableExtra("accessoryInfo");
        int vendorId = hidAccessoryInfo.getVendorId();
        int productId = hidAccessoryInfo.getProductId();
        String serialNumber = hidAccessoryInfo.getSerialNumber();
        RegistrationType registrationType = hidAccessoryInfo.getRegistrationType();
        String timeStamp = intent.getStringExtra("timestamp");
        EventCode eventCode = (EventCode) intent.getSerializableExtra("accessoryContextEventCode");
        String accessoryContextId = intent.getStringExtra("accessoryContextId");

        Log.i(TAG, "vendorId : "+vendorId);
        Log.i(TAG, "productId : "+productId);
        Log.i(TAG, "serialNumber : "+serialNumber);
        Log.i(TAG, "RegistrationType : "+registrationType);
        Log.i(TAG, "eventCode : "+eventCode);
        Log.i(TAG, "timeStamp : "+timeStamp);
        Log.i(TAG, "accessoryContextId : "+accessoryContextId);


        if(eventCode == EventCode.CONTEXT_CREATED || eventCode == EventCode.CONTEXT_RESENT){
            if (isInitializedSDK == false) {
                Log.d(TAG, "The SDK has not been initialized yet.");
                return;
            }
            if (!TextUtils.isEmpty(accessoryContextId)) {
                openAccessory(accessoryContextId);
            }
        } else if(eventCode == EventCode.CONTEXT_REVOKED){
            Log.i(TAG, "Accessory context is revoked.");
        }
    }

    public void onChange(Intent intent) {
        Log.i(TAG, "ACCESSORY_CHANGE_ACTION");


        HIDAccessoryInfo hidAccessoryInfo = intent.getParcelableExtra("accessoryInfo");
        int vendorId = hidAccessoryInfo.getVendorId();
        int productId = hidAccessoryInfo.getProductId();
        String serialNumber = hidAccessoryInfo.getSerialNumber();


        HIDReportEventInfo hidReportEventInfo = intent.getParcelableExtra("hidReportEventInfo");
        long reportOrdinal = hidReportEventInfo.getOrdinal();
        String timeStamp = hidReportEventInfo.getTimestamp();
        List<HIDReport> reports = hidReportEventInfo.getReports();

        Log.i(TAG, "vendorId : "+vendorId);
        Log.i(TAG, "productId : "+productId);
        Log.i(TAG, "serialNumber : "+serialNumber);
        Log.i(TAG, "ordinal : "+reportOrdinal);
        Log.i(TAG, "timeStamp : "+timeStamp);
        Log.i(TAG, "reportCount : "+reports.size());

        for(int i=0; i<reports.size();i++){
            HIDReport hidReport = reports.get(i);
            byte[] data = hidReport.getData();
            Log.i(TAG, "report_"+(i+1)+" : "+ new String(Base64.encode(data,Base64.DEFAULT)));
            if(isHPCardReader(vendorId, productId) && new String(data).contains("\r")){
                Principal principal = getCurrentPrincipal();
                if (principal != null && !principal.isAuthenticated()) {
                    initiateSignIn();
                } else {
                    signOut();
                }
            }
        }
    }

    private boolean isHPCardReader(int vendorId, int productId) {
        return vendorId == 1008 && productId == 69;
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/AccessoryAgentSample/src/com/hp/workpath/sample/accessoryagentsample/receiver/AccessoryReceiver.kt
package com.hp.workpath.sample.accessoryagentsample.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.text.TextUtils
import android.util.Base64
import android.util.Log
import com.hp.workpath.api.Result
import com.hp.workpath.api.SsdkUnsupportedException
import com.hp.workpath.api.Workpath
import com.hp.workpath.api.access.AccessService
import com.hp.workpath.api.access.Principal
import com.hp.workpath.api.accessory.RegistrationType
import com.hp.workpath.api.accessory.hid.AccessoryService
import com.hp.workpath.api.accessory.hid.EventCode
import com.hp.workpath.api.accessory.hid.HIDAccessoryInfo
import com.hp.workpath.api.accessory.hid.HIDInfo
import com.hp.workpath.api.accessory.hid.HIDReport
import com.hp.workpath.api.accessory.hid.HIDReportEventInfo
import com.hp.workpath.sample.accessoryagentsample.Logger
import com.hp.workpath.sample.accessoryagentsample.MainActivity
import java.lang.ref.WeakReference

class AccessoryReceiver : BroadcastReceiver() {
    private var mContextRef: WeakReference<Context?>? = null
    private var appContext: Context? = null

    override fun onReceive(context: Context, intent: Intent) {
        appContext = context.applicationContext
        mContextRef = WeakReference(appContext)

        initService(Handler(), appContext)

        if (ACCESSORY_CONTEXT_CHANGE_ACTION == intent.action || ACCESSORY_CHANGE_ACTION == intent.action) {
            val action = intent.action

            when (action) {
                ACCESSORY_CONTEXT_CHANGE_ACTION -> onContextChange(intent)
                ACCESSORY_CHANGE_ACTION -> onChange(intent)
            }
        }
    }

    private fun initService(handler: Handler, context: Context?) {
        try {
            Log.i(TAG, "initService Workpath.getInstance().initialize")
            if (!isInitializedSDK) {
                context?.let { Workpath.getInstance().initialize(it) }
                isInitializedSDK = true
            }
        } catch (e: SsdkUnsupportedException) {
            Logger.showResult(null, e.message)
        }
    }

    private fun openAccessory(accessoryContextId: String) {
        Log.d(TAG, "Accessory opened: $accessoryContextId")
        val result = Result()
        try {
            appContext?.let { AccessoryService.open(it, accessoryContextId, result) }
            Logger.showResult(null, "AccessoryService.open", result)
        } catch (e: Exception) {
            Log.e(TAG, e.message, e)
        }

        try {
            val hidInfo = appContext?.let { AccessoryService.getInfo(it, accessoryContextId, result) }
            if (result.code == Result.RESULT_OK && hidInfo != null) {
                Log.d(TAG, "HID feature report length : ${hidInfo.featureReportLength}")
                Log.d(TAG, "HID input report length : ${hidInfo.inputReportLength}")
                Log.d(TAG, "HID output report length : ${hidInfo.outputReportLength}")
            }
        } catch (e: Exception) {
            Log.e(TAG, e.message, e)
        }

        try {
            appContext?.let { AccessoryService.startReading(it, accessoryContextId, result) }
            Logger.showResult(null, "AccessoryService.startReading", result)
        } catch (e: Exception) {
            Log.e(TAG, e.message, e)
        }
    }

    private fun initiateSignIn() {
        val result = Result()
        Log.d(TAG, "initiateSignIn")
        appContext?.let { AccessService.initiateSignIn(it, result) }
        Logger.showResult(null, "AccessService.initiateSignIn", result)
    }

    private fun signOut() {
        val result = Result()
        Log.d(TAG, "signOut")
        appContext?.let { AccessService.signOut(it, result) }
        Logger.showResult(null, "AccessService.signOut", result)
    }

    private fun getCurrentPrincipal(): Principal? {
        val result = Result()
        Log.d(TAG, "getCurrentPrincipal")
        val principal = appContext?.let { AccessService.getCurrentPrincipal(it, result) }
        Logger.showResult(null, "AccessService.getCurrentPrincipal", result)
        return if (result.code == Result.RESULT_OK) {
            principal
        } else {
            null
        }
    }

    private fun onContextChange(intent: Intent) {
        val hidAccessoryInfo = intent.getParcelableExtra<HIDAccessoryInfo>("accessoryInfo")
        val vendorId = hidAccessoryInfo?.vendorId ?: -1
        val productId = hidAccessoryInfo?.productId ?: -1
        val serialNumber = hidAccessoryInfo?.serialNumber
        val registrationType = hidAccessoryInfo?.registrationType
        val timeStamp = intent.getStringExtra("timestamp")
        val eventCode = intent.getSerializableExtra("accessoryContextEventCode") as? EventCode
        val accessoryContextId = intent.getStringExtra("accessoryContextId")

        Log.i(TAG, "vendorId : $vendorId")
        Log.i(TAG, "productId : $productId")
        Log.i(TAG, "serialNumber : $serialNumber")
        Log.i(TAG, "RegistrationType : $registrationType")
        Log.i(TAG, "eventCode : $eventCode")
        Log.i(TAG, "timeStamp : $timeStamp")
        Log.i(TAG, "accessoryContextId : $accessoryContextId")

        if (eventCode == EventCode.CONTEXT_CREATED || eventCode == EventCode.CONTEXT_RESENT) {
            if (!isInitializedSDK) {
                Log.d(TAG, "The SDK has not been initialized yet.")
                return
            }
            if (!TextUtils.isEmpty(accessoryContextId)) {
                accessoryContextId?.let { openAccessory(it) }
            }
        } else if (eventCode == EventCode.CONTEXT_REVOKED) {
            Log.i(TAG, "Accessory context is revoked.")
        }
    }

    private fun onChange(intent: Intent) {
        Log.i(TAG, "ACCESSORY_CHANGE_ACTION")

        val hidAccessoryInfo = intent.getParcelableExtra<HIDAccessoryInfo>("accessoryInfo")
        val vendorId = hidAccessoryInfo?.vendorId ?: -1
        val productId = hidAccessoryInfo?.productId ?: -1
        val serialNumber = hidAccessoryInfo?.serialNumber

        val hidReportEventInfo = intent.getParcelableExtra<HIDReportEventInfo>("hidReportEventInfo")
        val reportOrdinal = hidReportEventInfo?.ordinal ?: -1
        val timeStamp = hidReportEventInfo?.timestamp
        val reports = hidReportEventInfo?.reports ?: emptyList()

        Log.i(TAG, "vendorId : $vendorId")
        Log.i(TAG, "productId : $productId")
        Log.i(TAG, "serialNumber : $serialNumber")
        Log.i(TAG, "ordinal : $reportOrdinal")
        Log.i(TAG, "timeStamp : $timeStamp")
        Log.i(TAG, "reportCount : ${reports.size}")

        for (i in reports.indices) {
            val hidReport = reports[i]
            val data = hidReport.data
            Log.i(TAG, "report_${i + 1} : ${String(Base64.encode(data, Base64.DEFAULT))}")
            if (isHPCardReader(vendorId, productId) && String(data).contains("\r")) {
                val principal = getCurrentPrincipal()
                if (principal != null && !principal.isAuthenticated) {
                    initiateSignIn()
                } else {
                    signOut()
                }
            }
        }
    }

    private fun isHPCardReader(vendorId: Int, productId: Int): Boolean {
        return vendorId == 1008 && productId == 69
    }

    companion object {
        const val TAG = MainActivity.TAG + "Receiver"
        private const val ACCESSORY_CONTEXT_CHANGE_ACTION = "com.hp.workpath.api.accessory.ACCESSORY_CONTEXT_CHANGE_ACTION"
        private const val ACCESSORY_CHANGE_ACTION = "com.hp.workpath.api.accessory.ACCESSORY_CHANGE_ACTION"
        private var isInitializedSDK = false
    }
}